notebooks/03-query/11-joins.ipynb

```diff
@@ -0,0 +1,2475 @@
+   "metadata": {},
+   "source": [
+    "### Results by Join Strategy"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 22,
+   "id": "81807f87-cff5-4e1c-8dd7-cc26b2b169e6",
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "[{'query_type': 'simple_join',\n",
+       "  'results': 'Results = avg:1182.9867 ms   min:1093.643 ms  max:1455.442 ms'},\n",
+       " {'query_type': 'simple_join_lookup',\n",
+       "  'results': 'Results = avg:254.4789 ms   min:234.103 ms  max:449.486 ms'},\n",
+       " {'query_type': 'simple_prejoin',\n",
+       "  'results': 'Results = avg:128.40235 ms   min:122.407 ms  max:176.861 ms'},\n",
+       " {'query_type': 'subquery_join',\n",
+       "  'results': 'Results = avg:2959.44585 ms   min:2841.148 ms  max:3093.195 ms'},\n",
+       " {'query_type': 'subquery_join_lookup',\n",
+       "  'results': 'Results = avg:550.8509 ms   min:519.703 ms  max:608.721 ms'},\n",
+       " {'query_type': 'subquery_prejoin',\n",
+       "  'results': 'Results = avg:352.1546 ms   min:327.998 ms  max:471.693 ms'}]"
+      ]
+     },
+     "execution_count": 22,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "# review the results\n",
+    "sorted_list = sorted(join_results, key=lambda x: x['query_type'])\n",
+    "sorted_list"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "66fd0a23-cc14-4537-8965-5dab8eeb1b46",
+   "metadata": {},
+   "source": [
+    "In these results you can see the progression of response times. \n",
+    "- The use of lookups improves performance over joining to another datasource, but it requires enough Heap to fit all lookups on all historical and ingestion tasks (Peon) JVMs.\n",
+    "- The pre-joined table performs even better and has the added advantage of not requiring any additional Heap.\n",
+    "\n",
+    "If the data being joined changes over time and you wish to query with the latest values, consider a lookup. If the data does not change or the change is not desirable, then pre-join the data. The example of `age_group` as the joined property works well in the pre-join strategy because the user's age_group is important at the time they clicked, not when looking at old click rows based on their current age_group."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "46e2e6f6-86e1-48dd-a125-6c1920173d74",
+   "metadata": {},
+   "source": [
+    "## Asynchronous Queries on MSQ Engine\n",
+    "\n",
+    "The MSQ (Multi-stage Query Engine) engine works by shuffling rows between stages. All stages execute in tasks in parallel as much as they can.\n",
+    "![](assets/msq-engine.png)\n",
+    "\n",
+    "Typically, as you did in the ingestion above, the MSQ engine is used to ingest external data, but it can also be used to query Druid datasources directly. It has the advantage of being able to do either `broadcast` or `sortMerge` joins. \n",
```

Comment:
Thought: is there a way to make the headings about different types of JOIN (LEFT, OUTER, INNER...) and then to add the different execution methods and words of warning within them? I suspect that someone looking at this notebook would benefit from seeing familiar terms in the main nav of the notebook (ie, "How do I do a LEFT JOIN in Druid?") - currently the notebook seems to lead with implementation details first.

Reply:
That's in the docs, that's not what this notebook is about.  The fact that it is a LEFT or OUTER join is somewhat immaterial to the subject of how they run. That's really just about learning SQL. We can do another notebook with LEFT/RIGHT/OUTER/FULL. (not sure all are supported).